integration-tests/itest/deploy_test.go

```diff
@@ -0,0 +1,461 @@
+		default:
+			parts := whitespaces.Split(line, AllSubstrings)
+			if len(parts) == 7 {
+				appEntry := appListEntry{parts[0], parts[1], parts[2], parts[3] + parts[4] + parts[5] + parts[6]}
+
+				res = append(res, appEntry)
+			}
+		}
+	}
+	return res, nil
+}
+
+func ensureAppNotDeployed(app app) error {
+	apps, err := listApps()
+	if err != nil {
+		return err
+	}
+	found := false
+	for _, entry := range apps {
+		if entry.name == app.name {
+			fmt.Printf("This is the app you are looking for: [%s]", entry.name)
+			found = true
+		}
+	}
+	if found {
+		err := undeploy(app)
+		if err != nil {
+			return err
+		}
+		err = ensureNoPods(app)
+		return err
+	}
+	return nil
+}
+
+func undeploy(app app) error {
+	fmt.Printf("Issuing Undeploy of app [%s]\n", app.name)
+	cmd := exec.Command("kubectl", "cloudflow", "undeploy", app.name)
+	_, err := cmd.CombinedOutput()
+	return err
+}
+
+func ensureNoPods(app app) error {
+	fmt.Printf("Ensuring no pods for app [%s]\n", app.name)
+	sleepDuration, err := time.ParseDuration("1s")
+	if err != nil {
+		log.Fatal("duration gives error", err)
+		return err // pfff
+	}
+	time.Sleep(sleepDuration)
+	pods, err := getPods(app)
+	if err != nil {
+		log.Fatal("get Pods gives error", err)
+		return err // pfff^2
+	}
+	fmt.Printf("Initial pod count %d\n", len(pods))
+
+	for len(pods) > 0 {
+
+		time.Sleep(sleepDuration)
+		pods, err = getPods(app)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("...%d", len(pods))
+	}
+	return nil
+}
+
+func getPods(app app) (pods []podEntry, err error) {
+	cmd := exec.Command("kubectl", "get", "pods", "-n", app.name)
+	out, er := cmd.CombinedOutput()
+	if er != nil {
+		err = er
+		return
+	}
+	str := string(out)
+	splits := strings.Split(str, "\n")
+	whitespaces := regexp.MustCompile(`\s+`)
+	var res []podEntry
+	for i, line := range splits {
+		switch i {
+		case 0:
+			continue
+		default:
+			parts := whitespaces.Split(line, AllSubstrings)
+			if len(parts) == 5 {
+				podEntry := podEntry{parts[0], parts[1], parts[2], parts[3], parts[4]}
+				res = append(res, podEntry)
+			}
+		}
+	}
+	return res, nil
+}
+
+func checkStatusIs(app app, status string) (res string, err error) {
+	for {
+		appStatus, er := getStatus(app)
+		if er != nil {
+			err = er
+			return
+		}
+		allSame := true
+		for _, entry := range appStatus.streamletPods {
+			allSame = allSame && entry.status == status
+			if !allSame {
+				fmt.Printf("Entry [%s, %s] is not compliant with status [%s]", entry.streamlet, entry.status, status)
+				break
+			}
+		}
+		if allSame {
+			return status, nil
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func getStreamlets(appStatus appStatus) []string {
+	var res []string
+	for _, entry := range appStatus.streamletPods {
+		res = append(res, entry.streamlet)
+	}
+	return res
+}
+
+func getStatus(app app) (status appStatus, err error) {
+	cmd := exec.Command("kubectl", "cloudflow", "status", app.name)
+	out, er := cmd.CombinedOutput()
+	if er != nil {
+		err = er
```

Comment:
Not sure if this is a pattern (with named return variables) commonly used in Go, it adds one line for every error also.

Reply:
how do you return only an error in this case? 
I can't:
`return err // doesn't match signature`
`return nil, err` // nil is not of type appStatus`
What's the way to _GO_ ?

Reply:
You should initialize an empty instance of the class and the error, some people also chose to return a pointer, then you can `return nil, err` , but I recommend the former.

Reply:
Yikes. Is there some form of type ascription possible?

Reply:
Stop fighting it :)

Reply:
:smile: